### PR TITLE
Initial contribution from Kaleido

### PR DESCRIPTION
**Pull request by peterbroadhurst**
See the README.md in the change for information.

Leaving in draft for now as there are some additional parts of the initial contribution pending:
- [x] README detailing references / credits
- [x] secp256k1 cryptography wrapper (ISC License)
- [x] Keystore V3 implementation (Scrypt + pbkdf2 derived keys, with AES/128/CTR encryption)
- [x] RLP encoding / decoding
- [x] Ethereum Transaction Encoding/Signing
    - [x] Original
    - [x] EIP-155
    - [x] EIP-1559 
- [x] Contribution of an existing JSON/RPC server framework for `eth_sendTransaction` based signing
    - [x] Configuration and logging in FireFly Microservice framework
    - [x] HTTPS JSON/RPC server
    - [x] `eth_sendTransaction` intercept
    - [x]  Simple `eth_getTransactionCount` support for missing Nonce
    - [x] Address lookup to Keystore V3 file
    - [x] Batch JSON/RPC support

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+## Contributing
+
+We welcome contributions to the FireFly Project in many forms, and
+there's always plenty to do!
+
+Please visit the
+[contributors guide](https://hyperledger.github.io/firefly//contributors/contributors.html) in the
```
Comment (awrichar) [inline CONTRIBUTING.md:7]: Probably an issue with the core repo too, but this page doesn't appear to exist.

File: LICENSE [hunk 2]
```diff
@@ -30,7 +30,7 @@
       "Object" form shall mean any form resulting from mechanical
       transformation or translation of a Source form, including but
       not limited to compiled object code, generated documentation,
-      and conversions to other media types.
+      and conversions to other media ethtypes.
```
Comment (awrichar) [inline LICENSE:33]: Looks like a stray find-replace.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+go-mod-tidy: .ALWAYS
+		$(VGO) mod tidy
+build: firefly-signer
+.ALWAYS: ;
+clean:
+		$(VGO) clean
+deps:
+		$(VGO) get ./ffsigner
+swagger:
+		$(VGO) test ./internal/apiserver -timeout=10s -tags swagger
```
Comment (awrichar) [inline Makefile:48]: Does this repo have an API server/Swagger reference?

Comment (peterbroadhurst) [reply Makefile:48]: No, everything is just the JSON/RPC `POST` `/` endpoint. Will remove this from `Makefile`

File: CONTRIBUTING.md (hunk 1 above)
Comment (peterbroadhurst) [reply CONTRIBUTING.md:7]: Changing to https://hyperledger.github.io/firefly/contributors/

File: cmd/ffsigner.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/hyperledger/firefly/pkg/log"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var sigs = make(chan os.Signal, 1)
+
+var rootCmd = &cobra.Command{
+	Use:   "ffsigner",
+	Short: "Hyperledger FireFly Tranansaction Manager",
```
Comment (awrichar) [inline cmd/ffsigner.go:41]: Wrong name here

File: cmd/ffsigner.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	rootCmd.AddCommand(versionCommand())
+	rootCmd.AddCommand(configCommand())
+}
+
+func Execute() error {
+	return rootCmd.Execute()
+}
+
+func initConfig() {
+	// Read the configuration, and register our policy engines
```
Comment (awrichar) [inline cmd/ffsigner.go:61]: No "policy engines" here

File: pkg/ethtypes/hexbytes.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// HexBytesPlain is simple bytes that are JSON stored/retrieved as hex
+type HexBytesPlain []byte
+
+// HexBytes0xPrefix is simple bytes that can o
```
Comment (awrichar) [inline pkg/ethtypes/hexbytes.go:29]: Comment got cut off?

File: internal/signerconfig/signerconfig.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+var ffc = config.AddRootKey
+
+var (
+	// BackendChainID optionally set the Chain ID manually (usually queries network ID)
+	BackendChainID = ffc("backend.chainId")
+	// FileWalletEnabled if the Keystore V3 wallet is enabled
+	FileWalletEnabled = ffc("fileWallet.enabled")
+	// FileWalletPath the path of the Keystore V3 wallet path
+	FileWalletPath = ffc("fileWallet.path")
+	// FileWalletFilenamesWith0xPrefix extension to append to the "from" address string to find the file (see metadata section for file types). All filenames must be lower case on disk.
```
Comment (awrichar) [inline internal/signerconfig/signerconfig.go:35]: I think this description is wrong (actually configures whether or not to expect the 0x prefix on filenames)

File: pkg/ethsigner/wallet.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"context"
+
+	"github.com/hyperledger/firefly-signer/pkg/ethtypes"
+)
+
+// Wallet is the common interface can be implemented across wallet/signing capabilities
+type Wallet interface {
+	Sign(ctx context.Context, txn *Transaction, chainID int64) ([]byte, error)
+	// SignPrivateTxn(ctx context.Context, addr ethtypes.Address, ptx *Transaction, chainID int64) ([]byte, error)
```
Comment (awrichar) [inline pkg/ethsigner/wallet.go:28]: Planned for future?

**Review by awrichar — APPROVED** [review]
Initial contribution looks great!

File: pkg/ethsigner/wallet.go (hunk 8 above)
Comment (peterbroadhurst) [reply pkg/ethsigner/wallet.go:28]: Yeah, didn't quite get to it this round. Kaleido has code for this we can contribute.

File: internal/signerconfig/signerconfig.go (hunk 7 above)
Comment (peterbroadhurst) [reply internal/signerconfig/signerconfig.go:35]: sorry - my english was trying to say that, I'll try to simplify per your suggestion

Comment (peterbroadhurst) [reply internal/signerconfig/signerconfig.go:35]: ah, yeah this was copy/paste error

**Review by awrichar — APPROVED** [review]

**Pull request merged**